modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1702,264 +1719,549 @@ def tree_reduce(
+            Data types of the result. This is an optimization
+            because there are functions that always result in a particular data
+            type, and allows us to avoid (re)computing it.
+        copy_dtypes : bool, default: False
+            If True, the dtypes of the resulting dataframe are copied from the original,
+            and the ``dtypes`` argument is ignored.
 
         Returns
         -------
         PandasDataframe
-            A new PandasDataframe with the inferred schema.
+            New Modin DataFrame.
         """
-        # Compute dtypes on the specified columns, and then set those dtypes on a new frame
-        new_cols = self.take_2d_labels_or_positional(col_labels=col_labels)
-        new_cols_dtypes = new_cols.tree_reduce(0, pandas.DataFrame.infer_objects).dtypes
-        new_dtypes = self.dtypes.copy()
-        new_dtypes[col_labels] = new_cols_dtypes
-        return self.__constructor__(
+        if other is not None:
+            if not isinstance(other, list):
+                other = [other]
+            other = [o._partitions for o in other] if len(other) else None
+
+        if apply_indices is not None:
+            numeric_indices = self.axes[axis ^ 1].get_indexer_for(apply_indices)
+            apply_indices = self._get_dict_of_block_index(
+                axis ^ 1, numeric_indices
+            ).keys()
+
+        new_partitions = self._partition_mgr_cls.map_partitions_full_axis(
             self._partitions,
-            self._index_cache,
-            self._columns_cache,
-            self._row_lengths_cache,
-            self._column_widths_cache,
-            new_dtypes,
+            self._build_treereduce_func(axis, func),
+            axis=axis,
+            other_partitions=other,
+            keep_partitioning=True,
+            apply_indices=apply_indices,
+            enumerate_partitions=enumerate_partitions,
         )
+        # Index objects for new object creation. This is shorter than if..else
+        kw = self._make_init_labels_args(new_partitions, new_index, new_columns)
+        if copy_dtypes:
+            kw["dtypes"] = self._dtypes
+        elif isinstance(dtypes, type):
+            kw["dtypes"] = pandas.Series(
+                [np.dtype(dtypes)] * len(kw["columns"]), index=kw["columns"]
+            )
+        result = self.__constructor__(new_partitions, **kw)
+        if new_index is not None:
+            result.synchronize_labels(axis=0)
+        if new_columns is not None:
+            result.synchronize_labels(axis=1)
+        return result
 
-    def join(
+    @lazy_metadata_decorator(apply_axis="both")
+    def map_select_indices(
         self,
-        axis: Union[int, Axis],
-        condition: Callable,
-        other: ModinDataframe,
-        join_type: Union[str, JoinType],
+        func: Callable,
+        *,
+        axis: AxisInt,
+        other=None,
+        full_axis=False,
+        apply_indices=None,
+        numeric_indices=None,
```

Comment:
do we really want these two parameters to exist at the same time? we can easily end-up in an ambiguous situation with this set of parameters:
```python
md_df.map_select_indices(
    apply_indices=["a", "b"],
    numeric_indices=[1, 2, 3, 4, 5], ...
) # what's the method supposed to do?
```